### PR TITLE
extend: add on_os (for on_macos/on_linux)

### PR DESCRIPTION
**Pull request by SeekingMeaning**
- [x] Have you followed the guidelines in our [Contributing](https://github.com/Homebrew/brew/blob/HEAD/CONTRIBUTING.md) document?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/brew/pulls) for the same change?
- [x] Have you added an explanation of what your changes do and why you'd like us to include them?
- [ ] Have you written new tests for your changes? [Here's an example](https://github.com/Homebrew/brew/blob/HEAD/Library/Homebrew/test/PATH_spec.rb).
- [x] Have you successfully run `brew style` with your changes locally?
- [ ] Have you successfully run `brew tests` with your changes locally?
- [ ] Have you successfully run `brew man` locally and committed any changes?

-----

This PR adds a module named `OnOS` that can be `include`d in a class to define the methods `on_macos` and `on_linux`

This reduces code duplication and should allow #9421 and #9438 to each be simplified to a single `include` statement

### CODE REVIEW

File: Library/Homebrew/extend/os_specifiable.rb [hunk 1]
```diff
@@ -0,0 +1,22 @@
+# typed: true
+# frozen_string_literal: true
+
+module OSSpecifiable
```
Comment (MikeMcQuaid) [inline Library/Homebrew/extend/os_specifiable.rb:4]:
```suggestion
module OnOS
```

Thoughts?

**Review by MikeMcQuaid — APPROVED** [review]
One naming question/debate but: I love this.

File: Library/Homebrew/extend/os/linux/on_os.rb [hunk 2]
```diff
@@ -0,0 +1,12 @@
+# typed: true
+# frozen_string_literal: true
+
+module OnOS
+  undef on_linux
```
Comment (reitermarkus) [inline Library/Homebrew/extend/os/linux/on_os.rb:5]:
```suggestion
```

File: Library/Homebrew/extend/os/linux/on_os.rb [hunk 3]
```diff
@@ -0,0 +1,12 @@
...
+
+  def on_linux(&block)
+    raise "No block content defined for on_linux block" unless block
```
Comment (reitermarkus) [inline Library/Homebrew/extend/os/linux/on_os.rb:8]:
```suggestion
    super
```

File: Library/Homebrew/extend/os/mac/on_os.rb [hunk 4]
```diff
@@ -1,22 +1,12 @@
 # typed: true
 # frozen_string_literal: true
 
-class Formula
+module OnOS
   undef on_macos
```
Comment (reitermarkus) [inline Library/Homebrew/extend/os/mac/on_os.rb:5]:
```suggestion
```

File: Library/Homebrew/extend/os/mac/on_os.rb [hunk 5]
```diff
@@ -1,22 +1,12 @@
...
 
   def on_macos(&block)
     raise "No block content defined for on_macos block" unless block
```
Comment (reitermarkus) [inline Library/Homebrew/extend/os/mac/on_os.rb:8]:
```suggestion
    super
```

Comment (SeekingMeaning) [reply Library/Homebrew/extend/os/mac/on_os.rb:8]:
```
NoMethodError: super: no superclass method `on_macos' for #<Formulary::FormulaNamespace5894bafd9cdb7978f7041fbe11653e6e::Make:0x00007f8b938fce30>
```

**Review by reitermarkus — APPROVED** [review]

**Pull request merged**